py/dml/expr.py

```diff
@@ -228,6 +229,74 @@ def typecheck_inargs(site, args, inp, kind="function", known_arglen=None):
         else:
             raise EPTYPE(site, arg, rtype, pname, kind)
 
+def typecheck_inarg_inits(site, inits, inp, location, scope,
+                          kind="function", variadic=False,
+                          allow_undefined_args=False):
+    if (not variadic and len(inits) != len(inp)) or len(inits) < len(inp):
+        raise EARG(site, kind)
+
+    from .expr_util import coerce_if_eint
+    from .codegen import eval_initializer, codegen_expression, \
+                         codegen_expression_maybe_nonvalue
```

Comment:
should some of these be promoted to expr.py? the original purpose of the expr.py / ctree.py split was to avoid some cyclic imports, so good to keep down the number of cycles. But only if it doesn't come with a big cost.

Reply:
this is cyclical in codegen.py and expr.py, which is much harder to resolve.